### ISSUE TITLE
[StB] Fix Incorrectly Reported Color Bonuses

### ISSUE THREAD

**Issue by KirbyComment** [touhouworldcup/thprac#164]
There is a strange display bug in StB where the displayed Color Bonuses are wrong when taking the calculation of the photo into account.

![image](https://github.com/touhouworldcup/thprac/assets/129665085/3472c259-2918-4796-8a20-7c7974a06012)

Taking this example into account. The first photo gets Red Shot but the second photo does not by including one less bullet in the photo. The Red Shot Bonus is 300, which with multipliers taken into account (approximated because of lacking decimals in the multipliers) would equal a difference of _1,290 points_ (rounded up), yet the real difference is _19,530 points_. This would make the real value for the Red Shot (in 10-8 at least) 4,560 points (not taking the one extra bullet into account but the point is that the difference between the reported Base Point increase is staggeringly different from the real Base Point increase.

![image](https://github.com/touhouworldcup/thprac/assets/129665085/161c5905-ffd7-4e26-be60-87ce9bb93aed)

This bug presumably applies to all color bonuses. I remember this being true for the Colorful Shot in 6-8 as well so this discrepancy is probably present in all scenes. If this is all correct, there should be an option in thprac to fix this bug and let the game display the true bonus that is being applied to the Base Value instead of what is displays on screen currently.

**Comment by H-J-Granger** [comment]
There seems to be no bug in the color bonus calculation. The reason for the huge difference here is that there is a hidden multiplier (float32 at `[4c4e6c]+0x25718`) here. Its value is 10.0x in phase 2 of 10-8, so the red bonus seems to be ~10x higher in the comparison.

**Comment by KirbyComment** [comment]
What about 6-8? Is there a hidden multiplier there too? And would it be possible to have this hidden multiplier be made visible? Regardless what it shows on the screen isn't the whole picture in regards to calculating score

**Comment by H-J-Granger** [comment]
> What about 6-8? Is there a hidden multiplier there too? And would it be possible to have this hidden multiplier be made visible? Regardless what it shows on the screen isn't the whole picture in regards to calculating score

Yes. The hidden multiplier is 5x in every photo of 6-8, and it is 10x in photos 4 to 9 in 10-8. Apparently, the hidden multiplier exists in every photo in the game, and its value only depends on its index (i.e., which level it's in, how many successful photos have been taken before it). (Probably) ZUN forgetting to set the multiplier back to 1x in photos 8 to 9 in 10-8 is the cause of them having an extremely high score compared to photos 1-3.

I've been working on a DS-styled bonus display feature recently, and I plan to add this multiplier (if it's not 1x) to the bottom of the multiplier list.

**Comment by H-J-Granger** [comment]
By the way, the hidden multiplier can be less than 1x. For example, it's 0.4x in 4-4 and 4-6, and 0.5x in 4-9.

Maybe I should write a blog post explaining the scoring mechanics of StB in detail in the future.

**Comment by KirbyComment** [comment]
A post explaining it in detail would be very appreciated. Maribel's Danmaku Wiki could use this information at the least

**Comment by H-J-Granger** [comment]
> A post explaining it in detail would be very appreciated. Maribel's Danmaku Wiki could use this information at the least

I've posted an explanatory article on [a Nyasama thread](https://bbs.nyasama.com/thread-1869948-1-1.html) (in Chinese). The English version, however, may take a while since I would like to post it on my blog, which is currently under construction.

**Comment by KirbyComment** [comment]
> > A post explaining it in detail would be very appreciated. Maribel's Danmaku Wiki could use this information at the least
> 
> I've posted an explanatory article on [a Nyasama thread](https://bbs.nyasama.com/thread-1869948-1-1.html) (in Chinese). The English version, however, may take a while since I would like to post it on my blog, which is currently under construction.

Awesome! Feel free to let me know when there's an English blog post of your findings. They could be useful for the Danmaku Gameplay Wiki.
The feature currently has some bugs in them but I reported them in #404 .
